segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/filesystem/FileSystemStorage.java

```diff
@@ -0,0 +1,377 @@
+        return supplyAsync(streamSegmentName, () -> syncExists(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentHandle> openWrite(String streamSegmentName) {
+        return supplyAsync(streamSegmentName, () -> syncOpenWrite(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<SegmentProperties> create(String streamSegmentName, Duration timeout) {
+        return supplyAsync( streamSegmentName, () -> syncCreate(streamSegmentName));
+    }
+
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration
+            timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncWrite(handle, offset, data, length));
+    }
+
+    @Override
+    public CompletableFuture<Void> seal(SegmentHandle handle, Duration timeout) {
+        return supplyAsync(handle.getSegmentName(), () -> syncSeal(handle));
+    }
+
+    @Override
+    public CompletableFuture<Void> concat(SegmentHandle targetHandle, long offset, String sourceSegment, Duration
+            timeout) {
+            return supplyAsync(targetHandle.getSegmentName(),
+                    () -> syncConcat(targetHandle, offset, sourceSegment));
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(SegmentHandle handle, Duration timeout) {
+            return supplyAsync(handle.getSegmentName(), () -> syncDelete(handle));
+    }
+
+    //endregion
+
+    //region AutoClosable
+
+    @Override
+    public void close() {
+
+    }
+
+    //endregion
+
+    //region private sync implementation
+
+    @SneakyThrows(StreamSegmentNotExistsException.class)
+    private SegmentHandle syncOpenRead(String streamSegmentName) {
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        }
+
+        return FileSystemSegmentHandle.getReadHandle(streamSegmentName);
+    }
+
+    @SneakyThrows
+    private SegmentHandle syncOpenWrite(String streamSegmentName) {
+        Path path = Paths.get(config.getFilesystemRoot(), streamSegmentName);
+        if (!Files.exists(path)) {
+            throw new StreamSegmentNotExistsException(streamSegmentName);
+        } else if (Files.isWritable(path)) {
+            return FileSystemSegmentHandle.getWriteHandle(streamSegmentName);
+        } else {
+                return openRead(streamSegmentName).get();
+        }
+    }
+
+    @SneakyThrows(IOException.class)
+    private int syncRead(SegmentHandle handle, long offset, byte[] buffer, int bufferOffset, int length) {
+
+        Path path = Paths.get(config.getFilesystemRoot(), handle.getSegmentName());
+
+            if (Files.size(path) < offset) {
+                log.warn("Read called on segment {} at offset {}. The offset is beyond the current size of the file.",
```

Comment:
We should decide what to do in these cases and come up with guidelines. I've had issues in the past when the code was not logging any message and was throwing an exception, but upstream the exception was being swallowed, so it was hard to debug the problem.

I agree that the log warn here is not strictly necessary and in some sense even undesirable because it adds to the logging overhead. If we are correctly handling this exception upstream, which I haven't checked, then it is fine to not have it, but we need to be in this mode of checking that this is being properly handled.